src/cloudflare/internal/vectorize-api.ts

```diff
@@ -0,0 +1,203 @@
+          accept: "application/json",
+        },
+      }
+    );
+
+    return await toJson<VectorizeVectorMutation>(res);
+  }
+
+  public async upsert(
+    vectors: VectorizeVector[]
+  ): Promise<VectorizeVectorMutation> {
+    const res = await this._send(
+      Operation.VECTOR_UPSERT,
+      `indexes/${this.indexId}/upsert`,
+      {
+        method: "POST",
+        body: JSON.stringify({
+          vectors: vectors.map((vec) => ({
+            ...vec,
+            values: Array.isArray(vec.values)
+              ? vec.values
+              : Array.from(vec.values),
+          })),
+        }),
+        headers: {
+          "content-type": "application/json",
+          "cf-vector-search-dim-width": String(
+            vectors.length ? vectors[0]?.values?.length : 0
+          ),
+          "cf-vector-search-dim-height": String(vectors.length),
+          accept: "application/json",
+        },
+      }
+    );
+
+    return await toJson<VectorizeVectorMutation>(res);
+  }
+
+  public async getByIds(ids: string[]): Promise<VectorizeVector[]> {
+    const res = await this._send(
+      Operation.VECTOR_GET,
+      `indexes/${this.indexId}/getByIds`,
+      {
+        method: "POST",
+        body: JSON.stringify(ids),
+        headers: {
+          "content-type": "application/json",
+          accept: "application/json",
+        },
+      }
+    );
+
+    return await toJson<VectorizeVector[]>(res);
+  }
+
+  public async deleteByIds(ids: string[]): Promise<VectorizeVectorMutation> {
+    const res = await this._send(
+      Operation.VECTOR_DELETE,
+      `indexes/${this.indexId}/deleteByIds`,
+      {
+        method: "POST",
+        body: JSON.stringify(ids),
+        headers: {
+          "content-type": "application/json",
+          accept: "application/json",
+        },
+      }
+    );
+
+    return await toJson<VectorizeVectorMutation>(res);
+  }
+
+  private async _send(
+    operation: Operation,
+    endpoint: string,
+    init: RequestInit
+  ): Promise<Response> {
+    const res = await this.fetcher.fetch(
+      `http://vector-search/binding/${endpoint}`,
+      init
+    );
+    if (res.status !== 200) {
+      try {
+        const err = (await res.json()) as VectorizeError;
+        throw new Error(
+          `${Operation[operation]}_ERROR ${
+            typeof err.code === "number" ? ` (code = ${err.code})` : ""
+          }: ${err.error}`,
+          {
+            cause: new Error(err.error),
+          }
+        );
+      } catch (e) {
+        throw new Error(
+          `${Operation[operation]}_ERROR: Status + ${res.status}`,
+          {
+            cause: new Error(`Status ${res.status}`),
+          }
+        );
+      }
+    }
+
+    return res;
+  }
+}
+
+async function toJson<T = unknown>(response: Response): Promise<T> {
+  const body = await response.text();
+  try {
+    return JSON.parse(body) as T;
+  } catch (e) {
+    throw new Error(`Failed to parse body as JSON, got: ${body}`);
```

Comment:
Is there ever a case here where `body` is huge? Specifically just worried about trying to drop too much into the log here. Might be worth truncating `body` to some reasonable max value